mixed-content/generic/mixed-content-test-case.js

```diff
@@ -0,0 +1,224 @@
+    "worker-request": requestViaWorker,
+    "xhr-request": requestViaXhr,
+    "audio-tag": requestViaAudio,
+    "video-tag": requestViaVideo,
+    "picture-tag": requestViaPicture,
+    "object-tag": requestViaObject,
+    "link-css-tag": requestViaLinkStylesheet,
+    "link-prefetch-tag": requestViaLinkPrefetch
+  };
+
+  sanityChecker.checkScenario(scenario, resourceInvoker);
+
+  var contentType = {
+    "a-tag": "application/json",
+    "area-tag": "application/json",
+    "fetch-request": "application/json",
+    "form-tag": "text/html",
+    "iframe-tag": "text/html",
+    "img-tag":  "image/png",
+    "script-tag": "application/javascript",
+    "worker-request": "application/javascript",
+    "xhr-request": "application/json",
+    "audio-tag": "audio/mpeg",
+    "video-tag": "video/mp4",
+    "picture-tag": "image/png",
+    "object-tag": "text/html",
+    "link-css-tag": "text/css",
+    "link-prefetch-tag": "text/html"
+  };
+
+  var assertInvoker = {
+    blocked: function(response) {
+      // When the resource is blocked, it cannot tear down the key in the server
+      // store. Therefore we expect to get back the value we sent in our
+      // resource request announcement.
+      assert_equals(response.status, "blocked",
+                    "The resource request should be blocked.");
+    },
+    allowed: function(response) {
+      // If the resource is allowed then it will tear down the key in the server
+      // store. This means when we try to read it, we will get an error.
+      assert_equals(response.status, "allowed",
+                    "The resource request should be allowed.");
+    }
+  };
+
+  var assertResponse = assertInvoker[scenario.expectation];
+
+  function queryResource(estimatedResponseTime) {
+    var key = guid();
+    var value = guid();
+    var announceResourceRequestUrl = endpoint['same-origin'] +
+                                     "?action=put&key=" + key +
+                                     "&value=" + value;
+    var assertResourceRequestUrl = endpoint['same-origin'] +
+                                  "?action=take&key=" + key;
+    var resourceRequestUrl = endpoint[scenario.origin] + "?redirection=" +
+                             scenario.redirection + "&action=purge&key=" +
+                             key + "&content_type=" +
+                             contentType[scenario.subresource];
+
+    var requestResource = resourceInvoker[scenario.subresource];
+    var queriedObject;
+
+    var p = xhrRequest(announceResourceRequestUrl)
+      .then(function(response) {
+        // Send out the real resource request.
+        // This should tear down the key if it's not blocked.
+        queriedObject = requestResource(resourceRequestUrl);
+
+        // Reuse the element promise when provided.
+        if ("eventPromise" in queriedObject)
+          return queriedObject.eventPromise;
+        else
+          return timeoutPromise(estimatedResponseTime);
```

Comment:
This timeout business continues to worry me in terms of time spent in tests. It's worth trying out to see what the impact is, but I suspect we're going to have to cull the tests down to use it, especially if CSP does the same.